packages/backend-defaults/src/entrypoints/database/connectors/postgres.ts

```diff
@@ -121,6 +138,119 @@ export async function buildPgDatabaseConfig(
   };
 }
 
+export type AzureTokenCredentialConfig = {
+  /**
+   * How early before an access token expires to refresh it with a new one.
+   * Defaults to 5 minutes
+   * Supported formats:
+   * - A string in the format of '1d', '2 seconds' etc. as supported by the `ms`
+   *   library.
+   * - A standard ISO formatted duration string, e.g. 'P2DT6H' or 'PT1M'.
+   * - An object with individual units (in plural) as keys, e.g. `{ days: 2, hours: 6 }`.
+   */
+  tokenRenewalOffsetTime?: string | HumanDuration;
+} & (
+  | {
+      type: 'ClientSecretCredential';
+      clientId: string;
+      clientSecret: string;
+      tenantId: string;
+    }
+  | {
+      type: 'ManagedIdentityCredential';
+      /**
+       * The client ID of a user-assigned managed identity.
+       * If not provided, the system-assigned managed identity is used.
+       */
+      clientId?: string;
+    }
+  | {
+      type: undefined | 'DefaultAzureCredential';
+    }
+);
+
+export async function buildAzurePgConfig(config: any, sanitizedConfig: any) {
+  const {
+    DefaultAzureCredential,
+    ManagedIdentityCredential,
+    ClientSecretCredential,
+  } = require('@azure/identity');
+
+  // By default get a new token starting five minutes before the old one expires
+  let tokenRenewalOffsetMs = 300_000;
+  const configReader = new ConfigReader(config);
+  if (configReader.has('connection.tokenCredential.tokenRenewalOffsetTime')) {
+    tokenRenewalOffsetMs = durationToMilliseconds(
+      readDurationFromConfig(new ConfigReader(config), {
+        key: 'connection.tokenCredential.tokenRenewalOffsetTime',
+      }),
+    );
+  }
+
+  let credential: TokenCredential;
+  if (configReader.has('connection.tokenCredential')) {
+    const tokenCredentialConfig = configReader.get(
+      'connection.tokenCredential',
+    ) as AzureTokenCredentialConfig;
+    switch (tokenCredentialConfig.type) {
+      case 'ManagedIdentityCredential':
+        if (tokenCredentialConfig.clientId) {
+          // Use user-assigned managed identity
+          credential = new ManagedIdentityCredential(
+            tokenCredentialConfig.clientId,
+          );
+        } else {
+          // Use system-assigned managed identity
+          credential = new ManagedIdentityCredential();
+        }
+        break;
+      case 'ClientSecretCredential':
+        credential = new ClientSecretCredential(
+          tokenCredentialConfig.tenantId,
+          tokenCredentialConfig.clientId,
+          tokenCredentialConfig.clientSecret,
+        );
+        break;
+      case 'DefaultAzureCredential':
+      case undefined:
+        credential = new DefaultAzureCredential();
+        break;
+      default:
+        throw new Error(
+          `Unknown token credential type: ${
+            (tokenCredentialConfig as any).type
```

Comment:
```suggestion
            (tokenCredentialConfig as any)?.type
```

just defensively against null :)